recipe/generate_llvm.py

```diff
@@ -0,0 +1,371 @@
+        "defined": set(),
+        "undefined": set()
+    }
+    out = check_output([os.environ["NM"], lib]).decode()
+    lines = out.split("\n")[1:]
+    for line in lines:
+        if line.startswith(str(lib)) or line.endswith(":"):
+            continue
+        split = line.split()
+        if len(split) == 0:
+            continue
+        state = split[-2].lower()
+        symbol = split[-1]
+        if state in ('b', 't', 'd'):
+            lookup[lib.name]["defined"].add(symbol)
+        elif state == "u":
+            lookup[lib.name]["undefined"].add(symbol)
+
+# Resolve undefined symbols
+for lib in libs:
+    for possible_lib in libs:
+        if len(lookup[lib.name]["undefined"]) == 0:
+            break
+        undefined_len = len(lookup[lib.name]["undefined"])
+        lookup[lib.name]["undefined"] -=  lookup[possible_lib.name]["defined"]
+        if len(lookup[lib.name]["undefined"]) < undefined_len:
+            deps[lib.name].add(possible_lib.name)
+
+# Resolve transitive dependencies; by using this approach we also can handle
+# cyclic dependencies between libraries.
+deps_changed = True
+while deps_changed:
+    deps_changed = False
+    for lib in libs:
+        dep_len = len(deps[lib.name])
+        for dep in list(deps[lib.name]):
+            deps[lib.name] |= deps[dep]
+        if len(deps[lib.name]) != dep_len:
+            deps_changed = True
+
+# The casing in the bazel files is different than the actual libs, so look it up here.
+case_lookup = {}
+for lib in deps.keys():
+    case_lookup[lib[3:-2].lower()] = lib[3:-2]
+
+# Shared code between LLVM and MLIR bazel generation
+
+def empty(*args, **kwargs):
+    pass
+
+def cc_library(name, deps=None, *args, **kwargs):
+    global output
+    if deps is None:
+        deps = []
+    # Remove third-party dependencies
+    # filtered_deps = [dep for dep in deps if (not dep.startswith("@llvm_"))]
+    filtered_deps = [dep for dep in deps if dep[1:] in LLVM_LIBS]
+
+    output += f"""
+cc_library(
+    name = "{name}",
+    linkopts = {make_linkopts(name, deps)},
+    visibility = ["//visibility:public"],
+    deps = {filtered_deps},
+)"""
+
+def import_file(f):
+    p = Path(f)
+    target = Path("llvm-project") / SYMBOL.lower() / p
+    target.parent.mkdir(parents=True, exist_ok=True)
+    origin = PREFIX / p
+    copyfile(origin, target)
+
+def glob(search, **kwargs):
+    out = []
+    cwd = os.getcwd()
+    os.chdir(os.environ["PREFIX"])
+    for a in search:
+        out += py_glob(a)
+    os.chdir(cwd)
+    return out
+
+def exports_files(files):
+    global output
+    for f in files:
+        if f in ["LICENSE.TXT", "run_lit.sh", "include/llvm/Config/config.h.cmake", "include/llvm/Config/llvm-config.h.cmake", "include/llvm/Config/abi-breaking.h.cmake"]:
+            continue
+        import_file(f)
+    output += f"""
+exports_files(
+    {files}
+)
+"""
+
+
+# Generate LLVM build defintion
+
+PREFIX = Path(os.environ["PREFIX"])
+SYMBOL = "LLVM"
```

Comment:
Maybe I'm overlooking something, but I find this a strange variable name. In the loop above you're really resolving symbols from the `nm` output, but calling a variable containing the string `"LLLVM"` a symbol is confusing me.